Add generated ProtoJSON marshalers

htdvisser opened the pull request:
<!--
Thanks for submitting a pull request. Please fill the template below,
otherwise we will not be able to process this pull request.
-->

#### Summary
<!--
A short summary, referencing related issues:
Closes #0000, References #0000, etc.
-->

This pull request is part of https://github.com/TheThingsNetwork/lorawan-stack/issues/2798. It adds JSON marshalers and unmarshalers for our protos, generated by https://github.com/TheThingsIndustries/protoc-gen-go-json.

Actual usage is still disabled by default, but can be enabled by settting the `TTN_LW_EXP_JSONPLUGIN` environment variable to true.

#### Changes
<!-- What are the changes made in this pull request? -->

- Add dependency `github.com/TheThingsIndustries/protoc-gen-go-json` v1.1.0
- Generate JSON marshalers with `ghcr.io/thethingsindustries/protoc:3.9.1-gen-go-json-1.1.0`
	- Note: This uses an old version of `protoc` because of our Gogo usage
- Add `(thethings.json.message) = { wrapper: true }` options to all wrapper types and regenerate
- Add `(thethings.json.enum)` and `(thethings.json.enum_value)` options to all enums
- Implement `MarshalProtoJSON` and `UnmarshalProtoJSON` for all types used in `(gogoproto.customtype)`.

#### Testing

<!-- How did you verify that this change works? -->

- Updated `pkg/ttnpb/ttnpb_encoding_test.go`
- 🐒 with the Console

##### Regressions

<!-- Please indicate features that this change could affect and how that was tested. -->

Mistakes in `protoc-gen-go-json` or in the options in our `.proto` files could break compatibility of our JSON API. Therefore this is still behind a feature flag, so that we can test this on non-production deployments before fully making the switch.

#### Notes for Reviewers
<!--
NOTE: This section is optional.

Motivate briefly why it is implemented this way, if that deviates from the
implementation proposal in the referenced issues.
- How should your reviewers approach this pull request?
- @mention reviewers with special requests or questions for them
-->

https://github.com/TheThingsIndustries/protoc-gen-go-json was (or is being) reviewed by @adriansmares and @KrishnaIyer.

I've put this in the v3.15.0 milestone, but if there isn't enough time to review this, we can also postpone it to v3.15.1.

#### Checklist
<!-- Make sure that this pull request is complete. -->

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
	- Since this is an experimental feature behind a feature flag, I don't think we should mention it.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

htdvisser committed:
api: Add support for generated ProtoJSON marshalers

htdvisser committed:
api: Generate JSON marshalers for wrapper types

htdvisser committed:
api: Generate JSON marshalers for enums

htdvisser committed:
api: Implement ProtoJSON marshalers for custom types

htdvisser committed:
api: Add tests for ProtoJSON encoding

htdvisser committed:
all: Support indentation with jsonplugin marshaler

htdvisser committed:
all: Introduce feature flag for generated JSON marshalers

adriansmares commented on api/devicerepository.proto:
```diff
@@ -58,25 +59,29 @@ message EndDeviceBrand {
+  KEY_PROVISIONING_CUSTOM = 1 [(thethings.json.enum_value) = { value: "custom" }];
   // Key Provisioning from the Global Join Server.
-  KEY_PROVISIONING_JOIN_SERVER = 2;
+  KEY_PROVISIONING_JOIN_SERVER = 2 [(thethings.json.enum_value) = { value: "join server" }];
```
Note to self: Yes, we did have spaces in the textual marshallers.

adriansmares commented on api/third_party/github.com/TheThingsIndustries/protoc-gen-go-json/annotations.proto:
```diff
@@ -0,0 +1,103 @@
+import "google/protobuf/descriptor.proto";
+
+// TODO: The extension number 1885 is temporary.
+// It will change after we register an extension number on
```
Is there an issue for this ?

htdvisser replied:
I've changed the number to 51885, which is in the range of numbers allocated for internal use in individual organizations. That way we don't need to register it.

Using 51885 doesn't prevent other projects from using it, as long as it doesn't clash with other extensions.

adriansmares commented on pkg/jsonpb/ttn.go:
```diff
@@ -14,25 +14,124 @@
+func (*TTNMarshaler) ContentType() string { return "application/json" }
+
+func (m *TTNMarshaler) Marshal(v interface{}) ([]byte, error) {
+	if marshaler, ok := v.(jsonplugin.Marshaler); ok && jsonpluginFeatureFlag {
```
Nit: I would wrap everything with the feature flag, to avoid the cast completely. Same bellow.

adriansmares commented:
On a second thought: what if we have another feature flag that allows us to randomly (let's say in 1% of the cases) uses both marshallers and logs an error if the values differ ?

The diff could be obtained by unmarshalling with the `encoding/json` or the `jsoniter` package to `map[string]interface{}`.

johanstokking commented on pkg/types/devaddr.go:
```diff
@@ -71,6 +72,29 @@ func (addr *DevAddr) UnmarshalJSON(data []byte) error {
+// MarshalProtoJSON implements the jsonplugin.Marshaler interface.
+func (addr *DevAddr) MarshalProtoJSON(s *jsonplugin.MarshalState) {
+	if addr == nil {
+		s.WriteNil()
```
return?

johanstokking replied:
Many of these occurrences

johanstokking reviewed (commented):
Looks very elegant, I had expected much worse. Great stuff!

htdvisser committed:
api: Upgrade protoc-gen-go-json to v1.1.0

htdvisser committed:
Merge v3.15 from TheThingsNetwork/lorawan-stack

htdvisser committed:
all: Wrap entire blocks with jsonplugin feature flag

htdvisser committed:
api: Return from MarshalProtoJSON after writing nil

htdvisser commented:
@adriansmares @johanstokking I've addressed your comments. Please take another look.

I'd like to get this in for v3.15.1 so that we can start testing it (using the feature flag).
